### PR TITLE
fix: hauling not getting disabled automatically with liquids

### LINKED ISSUE #3316: Hauling liquids causes hauling mode to not turn off automatically
### Describe the bug

Attempting to haul water activates hauling, but it does not turn off when you fail to pick up a liquid, and go into empty tiles

### Steps To Reproduce

1. Break Toilet, activate hauling.
2. Move around

### Screenshots

image

### Versions and configuration

- OS: Windows
    - OS Version: Windows 10 Pro for workstations
- Game Version: cd7556b [64-bit]
- Graphics Version: Tiles
- LAPI Version: 1
- Game Language: English [en_US]
- Mods loaded: [
    Bright Nights [bn],
    Elevated bridges [elevated_bridges],
    Craftable Gun Pack [craftgp],
    Alternative Map Key [alt_map_key],
    Fuji's Military Profession Pack [fuji_mpp],
    Arcana and Magic Items [Arcana],
    Craftable Bullet Casings [CasingCrafting],
    MST Extra [MST_Extra],
    Tankmod: Revived [Tankmod_Revived],
    Essence 2200 [Essence2200],
    Essence 2200 Tweaks [FoxTweaks],
    Nonperishable Overhaul [Nonperishable_Overhaul],
    No Addictions [No_Addictions],
    Limit Fungal Growth [limit_fungal_growth],
    Disable NPC Needs [no_npc_food]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by CrocodileJonesy**
## Summary

SUMMARY: Bugfixes "hauling not getting disabled automatically with liquids"

## Purpose of change

Fix for issue #3316, that is largely based on DDA code as well as porting

## Describe the solution

Fixes the bug  with all use cases I can think of by adding player hauling checks in code where hauling would normally be stopped

## Describe alternatives you've considered

None

## Testing

Tested use cases, I can think of, and all appear to work as intended without effecting anything else
Use cases that fix changes:
- Pressing haul key when only on spilt liquids
- Moving onto fences or obstacles such as windows etc.
- Diving into deep water
- When flung, grabbed, grappled
- Moving hauled items via advanced inventory manager

### Testing hauling with obstacles

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/51826149/6287b98e-b8a0-4ae8-9f23-3135bd3fdd75

### Testing hauling with spilt liquids

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/51826149/57611016-42a8-40c5-8d4a-c3ea8d866d21

### Testing with monsters attacking/interacting with player while hauling

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/51826149/f6095c9e-575c-40bc-87dd-f666032ec68d

### CODE REVIEW

File: src/map.cpp
Comment (scarf005) [inline src/map.cpp:1]: could we make `map::has_haulable_items` free function (not a member function of map)?
there's serveral benefits to not adding member functions:

1. The function does not modify anything but finds out whether an item is haulable or not in given position. 
2. The function does not require internal knowledge of map as `map::i_at` is public member function.
3. Almost every source file depends on `map.h`. Changing its signature causes significant part of the code to be rebuilt again, which may be time-consuming.

given 1,2 and 3, i think they would be better be contained in separate source file and header, for example `item_hauling.{cpp,h}`.

**Review by scarf005 — CHANGES_REQUESTED** [review]

File: src/map.cpp
Comment (CrocodileJonesy) [reply src/map.cpp:1]: Sure thing, will implement asap

Comment (CrocodileJonesy) [reply src/map.cpp:1]: Implemented requested changes, moved the hauling code into a separate class and updated references to it

Comment (scarf005) [reply src/map.cpp:1]: i'm wondering the reason it's a class rather than free function?

Comment (CrocodileJonesy) [reply src/map.cpp:1]: Oh, my bad. I appeared to have misread/misunderstood the suggestion, I thought moving into own h/cpp was it's own class. Will make adjustments

File: msvc-full-features/Cataclysm-lib-vcpkg-static.vcxproj
Comment (scarf005) [inline msvc-full-features/Cataclysm-lib-vcpkg-static.vcxproj:1]: would this be neccesary for building in msvc?

Comment (CrocodileJonesy) [reply msvc-full-features/Cataclysm-lib-vcpkg-static.vcxproj:1]: Ah, my bad again, seems VS was just being a bit silly, I reverted the change to the project file

**Review by scarf005 — APPROVED** [review]

**Pull request merged**